src/Sylius/Bundle/CoreBundle/CatalogPromotion/CommandHandler/RemoveInactiveCatalogPromotionHandler.php

```diff
@@ -0,0 +1,58 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\CoreBundle\CatalogPromotion\CommandHandler;
+
+use Doctrine\ORM\EntityManager;
+use SM\Factory\FactoryInterface;
+use SM\StateMachine\StateMachine;
+use Sylius\Bundle\CoreBundle\CatalogPromotion\Command\RemoveInactiveCatalogPromotion;
+use Sylius\Bundle\CoreBundle\CatalogPromotion\Exception\InvalidCatalogPromotionStateException;
+use Sylius\Component\Core\Model\CatalogPromotionInterface;
+use Sylius\Component\Promotion\Model\CatalogPromotionStates;
+use Sylius\Component\Promotion\Model\CatalogPromotionTransitions;
+use Sylius\Component\Promotion\Repository\CatalogPromotionRepositoryInterface;
+
+final class RemoveInactiveCatalogPromotionHandler
+{
+    public function __construct(
+        private CatalogPromotionRepositoryInterface $catalogPromotionRepository,
+        private EntityManager $entityManager,
+        private FactoryInterface $stateMachineFactory,
+    ) {
+    }
+
+    public function __invoke(RemoveInactiveCatalogPromotion $command): void
+    {
+        /** @var CatalogPromotionInterface|null $catalogPromotion */
+        $catalogPromotion = $this->catalogPromotionRepository->findOneBy(['code' => $command->code]);
```

Comment:
Maybe you can add state to the criteria not to check that later?
```suggestion
        $catalogPromotion = $this->catalogPromotionRepository->findOneBy(['code' => $command->code, 'state' => CatalogPromotionStates::STATE_INACTIVE]);
```

Reply:
First I did it this way, but there is a possibility ( in `async` processing with multiple workers ) that
 ```
$this->catalogPromotionRepository->findOneBy(['code' => $command->code, 'state' => CatalogPromotionStates::STATE_INACTIVE]);
```
returns `null` but in fact, this promotion is still being processed and we end up with an unremoved promotion.
Throwing an exception gives us another three ( by default ) attempts to process that message once other processes have finished price recalculating